source/val/validate_constants.cpp

```diff
@@ -0,0 +1,346 @@
+// Copyright (c) 2018 Google LLC.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/val/validate.h"
+
+#include "source/opcode.h"
+#include "source/val/instruction.h"
+#include "source/val/validation_state.h"
+
+namespace spvtools {
+namespace val {
+namespace {
+
+spv_result_t ValidateConstantBool(ValidationState_t& _,
+                                  const Instruction* inst) {
+  auto type = _.FindDef(inst->type_id());
+  if (!type || type->opcode() != SpvOpTypeBool) {
+    return _.diag(SPV_ERROR_INVALID_ID, inst)
+           << "Op" << spvOpcodeString(inst->opcode()) << " Result Type <id> '"
+           << _.getIdName(inst->type_id()) << "' is not a boolean type.";
+  }
+
+  return SPV_SUCCESS;
+}
+
+spv_result_t ValidateConstantComposite(ValidationState_t& _,
+                                       const Instruction* inst) {
+  std::string opcode_name = std::string("Op") + spvOpcodeString(inst->opcode());
```

Comment:
Driveby: To avoid always-constructing and appending the std::string, I'd be tempted to make this a lambda to be evaluated only when needed.  I have no idea if that would be faster.